kicker/main.c

```diff
@@ -1,302 +1,403 @@
+        charge_commanded_ = !charge_commanded_;
 
-    // Don't run the adc every loop
-    // 1000 * 10 us = 10 ms
-    if (time % 1000 == 0) {
-        current_voltage = get_voltage();
+      kick_db_down_ = kick_db_pressed;
+      chip_db_down_ = chip_db_pressed;
+      charge_db_down_ = charge_db_pressed;
+    } else {
+      /* PORTC &= ~(_BV(MCU_YELLOW)); */
     }
-    time++;
-}
-
-void update_ball_sense() {
-    static uint32_t ball_sense_change_count = 0;
 
-    // Filter ball value
-    // Want X amount in a row to be the same
-    bool new_reading = HAL_IsSet(BALL_SENSE_RX);
+    if (PINA & _BV(BALL_SENSE_RX))
+      PORTB &= ~(_BV(BALL_SENSE_LED));
+    else
+      PORTB |= _BV(BALL_SENSE_LED);
 
-    // If we sensed the ball, but don't think it's triggered
-    // or we didn't sense the ball, and think it's triggered
-    // AKA our estimate is wrong
-    if (ball_sensed ^ new_reading)
-        ball_sense_change_count++;
-    else // else correct reading
-        ball_sense_change_count = 0;
+    if (is_kicking())
+      PORTC &= ~(_BV(MCU_YELLOW));
+    else
+      PORTC |= _BV(MCU_YELLOW);
+    // get a voltage reading by weighing in a new reading, same concept as
+    // TCP RTT estimates (exponentially weighted sum)
 
-    // We got a wrong reading X times in a row
-    // so we should swap
-    if (ball_sense_change_count > BALL_SENSE_MAX_SAMPLES) {
-        ball_sense_change_count = 0;
-
-        ball_sensed = !ball_sensed;
+    // don't run the adc every loop
+    if (time % 1000 == 0) {
+      int voltage_accum =
+        (255 - kalpha) * last_voltage_ + kalpha * (get_voltage());
+      last_voltage_ = voltage_accum / 255;
+
+      int num_lights = ((int) last_voltage_ / 47);
+
+      PORTA |= (0x1F << 1);
+      //PORTA &= ~((0xFF >> abs(num_lights - sizeof(unsigned char) * 8)) << 1);
+      //PORTA &= 0b11011111;
+      switch (num_lights) {
+        case 1:
+          PORTA &= 0b11011111;
+          break;
+        case 2:
+          PORTA &= 0b11001111;
+          break;
+        case 3:
+          PORTA &= 0b11000111;
+          break;
+        case 4:
+          PORTA &= 0b11000011;
+          break;
+        case 5:
+          PORTA &= 0b11000001;
+          break;
+        default:
+          PORTA &= 0xFF;
+      }
     }
-}
+    time++;
 
-void charge_caps() {
     // if we dropped below acceptable voltage, then this will catch it
     // note: these aren't true voltages, just ADC output, but it matches
     // fairly close
-    
-    // Stop charging if we are at the voltage target
-    if (current_voltage > 244 ||
-        !charge_allowed ||
-        !command.commanded_charge) {
 
-        HAL_ClearPin(LT_CHARGE);
+    if (!(PIND & _BV(LT_DONE_N)) || last_voltage_ > 239 || !charge_allowed_ ||
+        !charge_commanded_) {
+      PORTD &= ~(_BV(LT_CHARGE));
+    } else if (last_voltage_ < 232 && charge_allowed_ && charge_commanded_) {
+      PORTD |= _BV(LT_CHARGE);
+    }
```

Comment:
There's a lot of terms here and it's very important this is correct, could this get commented?